### PR TITLE
ocm-minikube.sh create

### PR DESCRIPTION
**Pull request by hatfieldbrian**
ocm-minikube.sh automates @ShyamsundarR's [instructions](https://github.com/ShyamsundarR/ocm-minikube#readme) for creating an open-cluster-management (ocm) hub cluster, `hub`, that manages itself and a second cluster, `cluster1`.

Each cluster is a virtual cluster created locally using minikube and kvm.
Kvm installation is a prerequisite.
Platform is assumed to be Linux amd64.

### CODE REVIEW

**Comment by ShyamsundarR** [comment]
Tested this and works well. Just a single nit that can be dealt with later if this works as expected for others as well.

- (nit) Sleep time can be increased to say 5 seconds to reduce log noise

Faced this error once, completely unrelated to the script, noting here in case others hit it (failed to download operator-sdk):
```
+ make deploy-hub
Installing operator-sdk into '_output/tools/bin/operator-sdk'
mkdir -p '_output/tools/bin/'
curl -s -f -L https://github.com/operator-framework/operator-sdk/releases/download/v1.1.0/operator-sdk-v1.1.0-x86_64-linux-gnu -o '_output/tools/bin/operator-sdk'
make: *** [Makefile:185: ensure-operator-sdk] Error 28
```

**Comment by hatfieldbrian** [comment]
> * (nit) Sleep time can be increased to say 5 seconds to reduce log noise

Commit https://github.com/RamenDR/ramen/pull/46/commits/0e93934ff6b93853257587eb8c518424b2ac187c addresses excessive log noise from polling and preserves 1 second response time

File: hack/ocm-minikube.sh [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+if ! command -v minikube; then
+	# https://minikube.sigs.k8s.io/docs/start/
+	minikube_version=latest
+	minikube_version=v1.18.1
+	curl -LRo ${HOME}/.local/bin/minikube https://storage.googleapis.com/minikube/releases/${minikube_version}/minikube-linux-amd64
+	unset -v minikube_version
+	chmod +x ${HOME}/.local/bin/minikube
+fi
+
+minikube_start_options=--driver=kvm2
```
Comment (raghavendrabhat) [inline hack/ocm-minikube.sh:32]: According to [1] it makes use of some feature gates while starting the minikube cluster. Should they be considered as the minikube_start_options? 

======================================================================================

minikube start --force --memory="4096" --cpus="2" -b kubeadm --kubernetes-version="v1.19.2" --driver="kvm2" --feature-gates="BlockVolume=true,CSIBlockVolume=true,VolumeSnapshotDataSource=true,ExpandCSIVolumes=true" --profile="cluster1"

=========================================================================================

@ShyamsundarR 

[1] https://hackmd.io/239bv_V5QlaXBa1rl3aFAA?both

Comment (ShyamsundarR) [reply hack/ocm-minikube.sh:32]: > According to [1] it makes use of some feature gates while starting the minikube cluster. Should they be considered as the minikube_start_options?

No these are not required in this case, we should be fine.

Comment (Madhu-1) [reply hack/ocm-minikube.sh:32]: Most of these feature gates are already beta and enabled by default in latest kubernetes version.

**Comment by ShyamsundarR** [comment]
Is there anything pending on this? Can we get a second ack to get this merged?

**Review by ShyamsundarR — APPROVED** [review]

**Review by Madhu-1 — APPROVED** [review]

**Comment by vdeenadh** [comment]
I tried the script on Ubuntu 20.04 (kvm/amd64) and ran into one error or another:

Errors during first attempt:
```
FATA[0018] Failed to install OLM version "0.16.1": failed to create CRDs and resources: no matches for kind "OperatorGroup" in version "operators.coreos.com/v1" 
make: *** [Makefile:101: install-olm] Error 1
+ kubectl --context hub -n olm wait deployments --all --for condition=available --timeout 1m
timed out waiting for the condition on deployments/catalog-operator
timed out waiting for the condition on deployments/olm-operator
```
Errors during second attempt:
```
FATA[0592] Failed to run packagemanifests: error waiting for CSV to install: csv failed: reason: "InstallCheckFailed", message: "install timeout" 
make: *** [Makefile:112: deploy-hub-operator] Error 1
:
+ kubectl --context hub -n open-cluster-management wait deployments --all --for condition=available
deployment.apps/cluster-manager-registry-server condition met
error: timed out waiting for the condition on deployments/cluster-manager
+ echo exit value: 1
```
Third try:
```
🐳  Preparing Kubernetes v1.20.2 on Docker 20.10.3 ...
🔎  Verifying Kubernetes components...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v4
🌟  Enabled addons: default-storageclass, storage-provisioner
❌  Problems detected in kubelet:
    Apr 28 20:38:32 hub kubelet[4400]: E0428 20:38:32.637474    4400 pod_workers.go:191] Error syncing pod 4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc ("cluster-manager-5b4997789c-vvvzr_open-cluster-management(4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc)"), skipping: failed to "StartContainer" for "registration-operator" with CrashLoopBackOff: "back-off 5m0s restarting failed container=registration-operator pod=cluster-manager-5b4997789c-vvvzr_open-cluster-management(4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc)"
:
    Apr 28 20:40:42 hub kubelet[4400]: E0428 20:40:42.151611    4400 pod_workers.go:191] Error syncing pod 4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc ("cluster-manager-5b4997789c-vvvzr_open-cluster-management(4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc)"), skipping: failed to "StartContainer" for "registration-operator" with CrashLoopBackOff: "back-off 5m0s restarting failed container=registration-operator pod=cluster-manager-5b4997789c-vvvzr_open-cluster-management(4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc)"

❌  Exiting due to GUEST_START: wait 6m0s for node: wait for healthy API server: controlPlane never updated to v1.20.2
```

Fourth try:
```
Apr 28 20:48:48 hub kubelet[4400]: E0428 20:48:48.352723    4400 pod_workers.go:191] Error syncing pod 4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc ("cluster-manager-5b4997789c-vvvzr_open-cluster-management(4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc)"), skipping: failed to "StartContainer" for "registration-operator" with RunContainerError: "context deadline exceeded"
    Apr 28 20:54:29 hub kubelet[4400]: E0428 20:54:29.217731    4400 pod_workers.go:191] Error syncing pod 4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc ("cluster-manager-5b4997789c-vvvzr_open-cluster-management(4a2a4d2d-c3fe-4ca0-ab27-a03f72a99bcc)"), skipping: failed to "StartContainer" for "registration-operator" with RunContainerError: "context deadline exceeded"
:surfer:  Done! kubectl is now configured to use "hub" cluster and "default" namespace by default
mkdir -p munge-csv
cp deploy/cluster-manager/olm-catalog/cluster-manager/manifests/cluster-manager.clusterserviceversion.yaml munge-csv/cluster-manager.clusterserviceversion.yaml.unmunged
sed -e "s,quay.io/open-cluster-management/registration-operator:latest,quay.io/open-cluster-management/registration-operator:latest," -i deploy/cluster-manager/olm-catalog/cluster-manager/manifests/cluster-manager.clusterserviceversion.yaml
_output/tools/bin/operator-sdk run packagemanifests deploy/cluster-manager/olm-catalog/cluster-manager/ --namespace open-cluster-management --version 0.4.0 --install-mode OwnNamespace --timeout=10m
FATA[0002] Failed to run packagemanifests: create catalog: error creating catalog source: catalogsources.operators.coreos.com "cluster-manager-catalog" already exists
make: *** [Makefile:112: deploy-hub-operator] Error 1
```

**Comment by Madhu-1** [comment]
Opening a separate issue would be good to track it :)

**Pull request merged**